File: mini-services/agent-fleet/llm_backend.ts
/**
 * DenialDefender Gemini-Only LLM System — TypeScript/Bun
 * Backend: Google Gemini 3.5+ (direct API)
 *
 * Architecture:
 * - Uses Gemini direct API as the sole LLM backend
 * - No fallback — if Gemini fails, returns error LLMResponse
 * - Maintains consistent interface for all 8 agents
 */

// ─── Configuration from Environment ─────────────────────────────────────

const GEMINI_API_KEY = process.env.GEMINI_API_KEY ?? "";
const GEMINI_MODEL = process.env.GEMINI_MODEL ?? "gemini-3.6-flash";
const GEMINI_CHECK_TIMEOUT = parseInt(process.env.GEMINI_CHECK_TIMEOUT ?? "5", 10) * 1000;
const LLM_GENERATION_TIMEOUT = parseInt(process.env.LLM_GENERATION_TIMEOUT ?? "60", 10) * 1000;

// ─── Types ──────────────────────────────────────────────────────────────

enum LLMBackend {
  GEMINI_DIRECT = "gemini_direct",
}

interface LLMResponse {
  content: string;
  model: string;
  backend: LLMBackend;
  tokensUsed: number;
  success: boolean;
  error?: string;
}

interface GenerationConfig {
  temperature?: number;
  maxTokens?: number;
  model?: string;
  systemPrompt?: string;
}

// ─── GeminiLLM Class ───────────────────────────────────────────────────

class GeminiLLM {
  public activeBackend: LLMBackend = LLMBackend.GEMINI_DIRECT;
  public geminiAvailable: boolean | null = null;

  constructor() {
    this.checkGemini();
  }

  /** Check if Gemini API is accessible. */
  private checkGemini(): void {
    if (!GEMINI_API_KEY) {
      this.geminiAvailable = false;
      console.log("[GeminiLLM] No GEMINI_API_KEY set — agents will use mock mode");
      return;
    }

    try {
      const url = `https://generativelanguage.googleapis.com/v1beta/models?key=${GEMINI_API_KEY}`;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), GEMINI_CHECK_TIMEOUT);

      fetch(url, { signal: controller.signal })
        .then((res) => {
          if (!res.ok) throw new Error(`HTTP ${res.status}`);
          clearTimeout(timeoutId);
          this.geminiAvailable = true;
          console.log(`[GeminiLLM] Active LLM backend: ${this.activeBackend}`);
          return true;
        })
        .catch((err) => {
          clearTimeout(timeoutId);
          this.handleGeminiCheckError(err);
          return false;
        });

      // Optimistically return true; if the check fails later,
      // the generate method will handle it
      this.geminiAvailable = true;
    } catch (e) {
      this.handleGeminiCheckError(e);
    }
  }

  private handleGeminiCheckError(e: unknown): void {
    const msg = String(e);
    if (msg.includes("location is not supported")) {
      console.warn("[GeminiLLM] Gemini API geo-blocked in this region");
    } else if (msg.includes("leaked")) {
      console.warn("[GeminiLLM] Gemini API key reported as leaked");
    } else if (msg.includes("PERMISSION_DENIED")) {
      console.warn("[GeminiLLM] Gemini API permission denied");
    } else {
      console.warn(`[GeminiLLM] Gemini API check failed: ${msg.slice(0, 100)}`);
    }
    this.geminiAvailable = false;
  }

  /** Generate text using Gemini direct API. No fallback. */
  async generate(
    prompt: string,
    config: GenerationConfig = {}
  ): Promise<LLMResponse> {
    const { temperature = 0.7, maxTokens = 2048, model, systemPrompt = "" } = config;
    return this.generateGemini(prompt, systemPrompt, model, temperature, maxTokens);
  }

  /** Generate using Gemini API directly. */
  private async generateGemini(
    prompt: string,
    systemPrompt: string,
    model?: string,
    temperature = 0.7,
    maxTokens = 2048
  ): Promise<LLMResponse> {
    try {
      const modelName = model ?? GEMINI_MODEL;
      const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelName}:generateContent?key=${GEMINI_API_KEY}`;

      const contents: Array<{ role: string; parts: Array<{ text: string }> }> = [];
      if (systemPrompt) {
        contents.push({ role: "user", parts: [{ text: systemPrompt }] });
        contents.push({ role: "model", parts: [{ text: "Understood." }] });
      }
      contents.push({ role: "user", parts: [{ text: prompt }] });

      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), LLM_GENERATION_TIMEOUT);

      const res = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          contents,
          generationConfig: {
            temperature,
            maxOutputTokens: maxTokens,
          },
        }),
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!res.ok) {
        const errText = await res.text();
        throw new Error(`HTTP ${res.status}: ${errText.slice(0, 200)}`);
      }

      const data = await res.json();
      const text = data.candidates[0].content.parts[0].text;
      const tokens = data.usageMetadata?.totalTokenCount ?? 0;

      return {
        content: text,
        model: modelName,
        backend: LLMBackend.GEMINI_DIRECT,
        tokensUsed: tokens,
        success: true,
      };
    } catch (e) {
      const error = String(e);
      console.error(`[GeminiLLM] Gemini API error: ${error.slice(0, 200)}`);
      return {
        content: "",
        model: model ?? GEMINI_MODEL,
        backend: LLMBackend.GEMINI_DIRECT,
        tokensUsed: 0,
        success: false,
        error,
      };
    }
  }
}

// ─── Singleton Instance ────────────────────────────────────────────────

let _llmInstance: GeminiLLM | null = null;

export function getLLM(): GeminiLLM {
  if (!_llmInstance) {
    _llmInstance = new GeminiLLM();
  }
  return _llmInstance;
}

export { GeminiLLM, LLMBackend };
export type { LLMResponse, GenerationConfig };

File: mini-services/agent-fleet/llm_backend.py
"""
DenialDefender Gemini-Only LLM System
Backend: Google Gemini 3.5+ (direct API)

Architecture:
- Uses Gemini direct API as the sole LLM backend
- No fallback — if Gemini fails, returns error LLMResponse
- Maintains consistent interface for all 8 agents
"""

import os
import json
import logging
from typing import Optional
from dataclasses import dataclass
from enum import Enum

logger = logging.getLogger(__name__)

class LLMBackend(Enum):
    GEMINI_DIRECT = "gemini_direct"

@dataclass
class LLMResponse:
    content: str
    model: str
    backend: LLMBackend
    tokens_used: int = 0
    success: bool = True
    error: Optional[str] = None

class GeminiLLM:
    """
    Gemini-only LLM client using Google Gemini 3.5+ direct API.
    No fallback backend — if Gemini fails, returns error LLMResponse.
    """
    
    def __init__(self):
        self.gemini_api_key = os.environ.get("GEMINI_API_KEY", "")
        self.gemini_model = os.environ.get("GEMINI_MODEL", "gemini-3.5-flash")
        self.active_backend: LLMBackend = LLMBackend.GEMINI_DIRECT
        self.gemini_available: Optional[bool] = None
        self._check_gemini()
    
    def _check_gemini(self) -> None:
        """Check if Gemini API is accessible."""
        if not self.gemini_api_key:
            self.gemini_available = False
            logger.info("No GEMINI_API_KEY set — agents will use mock mode")
            return
        
        try:
            import urllib.request
            url = f"https://generativelanguage.googleapis.com/v1beta/models?key={self.gemini_api_key}"
            req = urllib.request.Request(url)
            response = urllib.request.urlopen(req, timeout=5)
            self.gemini_available = True
        except Exception as e:
            error_msg = str(e)
            if "location is not supported" in error_msg:
                logger.warning("Gemini API geo-blocked in this region")
            elif "leaked" in error_msg:
                logger.warning("Gemini API key reported as leaked")
            elif "PERMISSION_DENIED" in error_msg:
                logger.warning("Gemini API permission denied")
            else:
                logger.warning(f"Gemini API check failed: {error_msg[:100]}")
            self.gemini_available = False
        
        if self.gemini_available:
            logger.info(f"Active LLM backend: {self.active_backend.value}")
        else:
            logger.warning("Gemini API not available — agents will use mock mode")
    
    def generate(self, prompt: str, system_prompt: str = "", 
                 model: Optional[str] = None, temperature: float = 0.7,
                 max_tokens: int = 2048) -> LLMResponse:
        """Generate text using Gemini direct API. No fallback."""
        return self._generate_gemini(prompt, system_prompt, model, temperature, max_tokens)
    
    def _generate_gemini(self, prompt: str, system_prompt: str,
                         model: Optional[str], temperature: float,
                         max_tokens: int) -> LLMResponse:
        """Generate using Gemini API directly."""
        try:
            import urllib.request
            model_name = model or self.gemini_model
            url = f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}:generateContent?key={self.gemini_api_key}"
            
            contents = []
            if system_prompt:
                contents.append({"role": "user", "parts": [{"text": system_prompt}]})
                contents.append({"role": "model", "parts": [{"text": "Understood."}]})
            contents.append({"role": "user", "parts": [{"text": prompt}]})
            
            body = json.dumps({
                "contents": contents,
                "generationConfig": {
                    "temperature": temperature,
                    "maxOutputTokens": max_tokens
                }
            }).encode()
            
            req = urllib.request.Request(url, data=body, method='POST')
            req.add_header('Content-Type', 'application/json')
            response = urllib.request.urlopen(req, timeout=60)
            data = json.loads(response.read().decode())
            
            text = data['candidates'][0]['content']['parts'][0]['text']
            tokens = data.get('usageMetadata', {}).get('totalTokenCount', 0)
            
            return LLMResponse(
                content=text,
                model=model_name,
                backend=LLMBackend.GEMINI_DIRECT,
                tokens_used=tokens,
                success=True
            )
        except Exception as e:
            error_msg = str(e)
            logger.error(f"Gemini API error: {error_msg[:200]}")
            return LLMResponse(
                content="",
                model=model or self.gemini_model,
                backend=LLMBackend.GEMINI_DIRECT,
                success=False,
                error=error_msg
            )

# Singleton instance
_llm_instance: Optional[GeminiLLM] = None

def get_llm() -> GeminiLLM:
    """Get the singleton LLM instance."""
    global _llm_instance
    if _llm_instance is None:
        _llm_instance = GeminiLLM()
    return _llm_instance
